fv/service_loop_prevention_test.go

```diff
@@ -116,13 +119,77 @@ var _ = infrastructure.DatastoreDescribe("service loop prevention; with 2 nodes"
 		}
 	}
 
+	tryRoutingLoop := func(expectLoop bool) {
+
+		// Run containers to model a default gateway, and an external client connecting to
+		// services within the cluster via that gateway.
+		externalGW := containers.Run("external-gw",
+			containers.RunOpts{AutoRemove: true},
+			"--privileged", // So that we can add routes inside the container.
+			utils.Config.BusyboxImage,
+			"/bin/sh", "-c", "sleep 1000")
+		defer externalGW.Stop()
+
+		externalClient := containers.Run("external-client",
+			containers.RunOpts{AutoRemove: true},
+			"--privileged", // So that we can add routes inside the container.
+			utils.Config.BusyboxImage,
+			"/bin/sh", "-c", "sleep 1000")
+		defer externalClient.Stop()
+
+		// Add a service CIDR route in those containers, similar to the routes that they
+		// would have via BGP per our service advertisement feature.  (This should really be
+		// an ECMP route to both Felixes, but busybox's ip can't program ECMP routes, and a
+		// non-ECMP route is sufficient to demonstrate the looping issue.)
+		externalClient.Exec("ip", "r", "a", "10.96.0.0/17", "via", externalGW.IP)
+		externalGW.Exec("ip", "r", "a", "10.96.0.0/17", "via", felixes[0].IP)
+
+		// Configure the external gateway client to forward, in order to create the
+		// conditions for looping.
+		externalClient.Exec("sysctl", "-w", "net.ipv4.ip_forward=1")
+		externalGW.Exec("sysctl", "-w", "net.ipv4.ip_forward=1")
+
+		// Also tell Felix to route that CIDR to the external gateway.
+		felixes[0].ExecMayFail("ip", "r", "d", "10.96.0.0/17")
+		felixes[0].Exec("ip", "r", "a", "10.96.0.0/17", "via", externalGW.IP)
+		felixes[0].Exec("iptables", "-P", "FORWARD", "ACCEPT")
+
+		tryPing := func() int {
+			// Start monitoring all packets, on the Felix, to or from a specific (but
+			// unused) service IP.
+			tcpdumpF := felixes[0].AttachTCPDump("eth0")
+			tcpdumpF.AddMatcher("serviceIPPackets", regexp.MustCompile("10\\.96\\.0\\.19"))
+			tcpdumpF.Start()
+			defer tcpdumpF.Stop()
+
+			// Send a single ping from the external client to the unused service IP.
+			err := externalClient.ExecMayFail("ping", "-c", "1", "-W", "1", "10.96.0.19")
+			Expect(err).To(HaveOccurred())
+
+			// Return the number of packets observed to/from 10.96.0.19.
+			return tcpdumpF.MatchCount("serviceIPPackets")
+		}
+
+		if expectLoop {
+			// Tcpdump should see more than 2 packets, because of looping.  Note: 2
+			// packets would be Felix receiving the ping and then forwarding it out
+			// again.  I want to check here that it's also looped around again by the
+			// gateway, resulting in MORE THAN 2 packets.
+			Expect(tryPing()).To(BeNumerically(">", 2))
```

Comment:
`MatchCount` is async (it depends on reading from the stdout pipe of tcpdump) so it should always be used with an `Eventually` (or a retry loop).

It's usually quite fast but I don't think it's guaranteed to work here.